src/raftstore/store/peer_storage.rs

```diff
@@ -308,29 +308,53 @@ impl InvokeContext {
     }
 
     #[inline]
-    pub fn save_raft_to(&self, db: &DB, wb: &mut WriteBatch) -> Result<()> {
-        let handle = try!(rocksdb::get_cf_handle(db, CF_RAFT));
-        try!(wb.put_msg_cf(handle,
-                           &keys::raft_state_key(self.region_id),
-                           &self.raft_state));
+    pub fn save_raft_to(&self, wb: &mut WriteBatch) -> Result<()> {
+        try!(wb.put_msg(&keys::raft_state_key(self.region_id), &self.raft_state));
         Ok(())
     }
 
     #[inline]
-    pub fn save_apply_to(&self, db: &DB, wb: &mut WriteBatch) -> Result<()> {
-        let handle = try!(rocksdb::get_cf_handle(db, CF_RAFT));
-        try!(wb.put_msg_cf(handle,
-                           &keys::apply_state_key(self.region_id),
-                           &self.apply_state));
+    pub fn save_apply_to(&self, wb: &mut WriteBatch) -> Result<()> {
+        try!(wb.put_msg(&keys::apply_state_key(self.region_id), &self.apply_state));
         Ok(())
     }
 }
 
-fn init_raft_state(engine: &DB, region: &Region) -> Result<RaftLocalState> {
-    Ok(match try!(engine.get_msg_cf(CF_RAFT, &keys::raft_state_key(region.get_id()))) {
+fn init_raft_state(engine: &DB, raft_engine: &DB, region: &Region) -> Result<RaftLocalState> {
+    let raft_state_key = keys::raft_state_key(region.get_id());
+    let region_state: Option<RegionLocalState> =
+        try!(engine.get_msg(&keys::region_state_key(region.get_id())));
+
+    // in case of restart happen when we just write region state to Applying,
+    // but not write raft_local_state to raft rocksdb in time.
+    if region_state.is_some() && region_state.as_ref().unwrap().get_state() == PeerState::Applying {
+        let raft_wb = WriteBatch::new();
+        box_try!(clear_raft_data(&raft_wb, raft_engine, region.get_id()));
+
+        let raft_state: RaftLocalState = match box_try!(engine.get_msg(&raft_state_key)) {
+            Some(state) => state,
+            None => {
+                return Err(box_err!("failed to get raftstate from {}", escape(&raft_state_key)))
+            }
+        };
+        try!(raft_wb.put_msg(&raft_state_key, &raft_state));
+        try!(raft_engine.write(raft_wb));
+        return Ok(raft_state);
+    }
+
+    Ok(match try!(raft_engine.get_msg(&raft_state_key)) {
         Some(s) => s,
         None => {
             let mut raft_state = RaftLocalState::new();
+            if region_state.is_some() && region_state.unwrap().get_state() == PeerState::Normal {
```

Comment:
you can use region.get_peers().is_empty() to check whether it is initalized or not.